rfcs/select-component.md

```diff
@@ -0,0 +1,191 @@
+  * `nextState` - a new state changes that will be set
+  * `currentState` - current full state of the component
+  * `params` may contain `id` and `label` of selected option and `selectedOptions` array of all of selected, as well as new `textValue` set in input.
+* `options: func` should return `Select` instance with standard or customized inner elements.
+* `onSelect: func(e: SyntheticInputEvent, params: Object)`:
+  handler for events on trigger element when option is selected. `params` has `id` and `label` of selected option and `selectedOptions` array of all of selected.
+* `onUnSelect: func(e: SyntheticInputEvent, params: Object)`:
+  handler for events on trigger element when option is unselected (for multiple choices). `params` has `id` and `label` of unSelected option and `selectedOptions` array of all of selected.
+* `onClearAll: func(e: SyntheticInputEvent, params: Object)`:
+  handler for events on trigger element when input is cleared (in Search mode). `params` has new `textValue` set in input.
+* `onKeyUp: func(e: SyntheticInputEvent, params: Object)`:
+  handler for events on trigger element when input has changed value (in Search mode). Handler for changing possible Options (as options). `params` has new `textValue` set in input.
+* `onMouseEnter: func`:
+  handler for events on trigger element
+* `onMouseLeave: func`:
+  handler for events on trigger element
+* `onFocus: func`:
+  handler for events on trigger element
+* `onBlur: func`:
+  handler for events on trigger element
+
+### Usage
+
+```js
+import {
+  StatefulSelect,
+  Select,
+  StyledRoot,
+  StyledInput,
+  StyledInputContainer,
+  StyledTag,
+  StyledSearchIcon,
+  StyledDropDown,
+  StyledOption,
+  ICON,
+  OPTIONS,
+  TYPE,
+} from './index';
+
+import {withStyle} from 'styletron-react';
+
+const CustomOption = withStyle(StyledOption, {
+  textColor: 'red',
+});
+
+export default () => {
+  const options = [
+    {
+      id: '1',
+      label: (
+        <span>
+          <img
+            style={{
+              borderRadius: '50%',
+              height: '75px',
+            }}
+            src="1.jpg"
+          />First
+        </span>
+      ),
+    },
+    {
+      id: '2',
+      disabled: true,
+      label: (
+        <span>
+          <img
+            style={{
+              borderRadius: '50%',
+              height: '75px',
+            }}
+            src="2.jpg"
+          />Second
+        </span>
+      ),
+    },
+  ];
+  return (
+    <React.Fragment>
+      <StatefulSelect
+        options={options}
+        initialState={{
+          selectedOptions: [
+            {
+              id: '3',
+              label: (
+                <span>
+                  <img
+                    style={{
+                      borderRadius: '50%',
+                      height: '75px',
+                    }}
+                    src="3.jpg"
+                  />Third
+                </span>
+              ),
+            },
+          ],
```

Comment:
Putting react components in state seems like an anti-pattern.
http://web.archive.org/web/20150419023006/http://facebook.github.io/react/docs/interactivity-and-dynamic-uis.html

 I think instead of allowing people to pass label as `React.Node`, they should just specify a component override for `Item` or we should expose a renderProp.

Reply:
`label` is whatever renderable is inside of Option. We do not even care, what's there giving total flexibility. So there is nothing to override, otherwise we shall have default layout for it, but we don't have it by design. The same approach is used in Menu, actually it is made for Menu as we re-use it.
Yes, state shouldn't contain elements, but from perspective of Select it is a data to be shown.
Being able to support a dozen different component in labels (especially for Search results) seems to be more better, than do this work in one `override` class function

Reply:
Does anyone have a use case for supporting a dozen different labels inside a Select? I'd argue they don't.

Additionally, if we want to support filtering the options based on keywords the user entered, this label needs to be a string.

Reply:
There are examples in internet. Not dozens, but a few. I can do as you wish. I can treat label as a `string` by default and if it is object containing data than customer should provide override to treat this data correspondingly.
![image](https://user-images.githubusercontent.com/1851469/44358815-edecae00-a46a-11e8-94b3-6ccf67e96545.png)

Reply:
Yeah the Amazon example is an interesting case. I'd probably still lean towards handling this via an override component or `renderOption` prop, but I agree it's not cut and dry.

Maybe we can get feedback from some more folks.

Reply:
I changed towards label to be pure string or object of data, in second case customer is obliged to provide custom code to return label layout based on label data